serverless/cmd/integrate/main.go

```diff
@@ -125,12 +157,179 @@ func main() {
 		glog.Exit("Nothing to integrate")
 	}
 
-	err = signAndWrite(ctx, newCp, cpNote, s, st)
+	err = signAndWrite(ctx, newCp, cpNote, signer, st)
 	if err != nil {
 		glog.Exitf("Failed to sign: %q", err)
 	}
 }
 
+// Extracts the key's name (for golang Note purposes) from the key metadata. It is an error if the name field
+// does not exist within the metadata. Note that this is separate from the key's actual name in DSM, which
+// must be unique (within the account).
+func getDsmKeyName(key *sdkms.Sobject) (string, error) {
+	if key.CustomMetadata == nil {
+		return "", fmt.Errorf("key had no custom metadata")
+	}
+
+	return (*key.CustomMetadata)["name"], nil
+}
+
+// Returns a Note.Verifier for a DSM signing key's public key. Since this is the public key, we can get the actual
+// key material and verify signatures locally without using DSM.
+func makeDsmVerifier(name string, key *sdkms.Sobject) (note.Verifier, uint32, error) {
+	pubAny, err := x509.ParsePKIXPublicKey(*key.PubKey)
+
+	pub := pubAny.(ed25519.PublicKey)
+	if len(pub) != Ed25519PubKeyLength {
+		return nil, 0,fmt.Errorf("Invalid key length for ed25519 public key. Expected %d, got %d",
+			Ed25519PubKeyLength, len(pub))
+	}
+
+	// In Note format, the first byte of the key is the key algorithm. 1 indicates ed25519, which is the only
+	// supported key type right now.
+	pub = append([]byte{1}, pub...)
+	
+	pubKeyHash := keyHash(name, pub)
+	
+	// Unfortunately, the only provided interface to construct a standard Note verifier is via this formatted
+	// string. The alternative here is creating our own Verifier-compatible type we could directly populate,
+	// but that would involve essentially copying code from the Note implementation. It's simpler (if uglier)
+	// to string-format the public key in the way expected by Note.
+	keyString := fmt.Sprintf("%s+%08x+%s", name, pubKeyHash, base64.StdEncoding.EncodeToString(pub))
+
+	// Write the public key to a file if requested.
+	if len(*pubKeyOut) > 0 {
+		err := writeFileIfNotExists(*pubKeyOut, keyString)
+		if err != nil {
+			return nil, 0, fmt.Errorf("Unable to write public key to %s: %q", *pubKeyOut, err)
+		}
+	}
+
+	verifier, err := note.NewVerifier(keyString)
+	if err != nil {
+		return nil, 0, fmt.Errorf("Unable to construct note.Verifier: %q", err)
+	}
+
+	return verifier, pubKeyHash, nil
+}
+
+// Create a Note.Signer-compatible signer that will sign using a DSM key. Unlike with the Verifier, we have to
+// construct our own Signer type that performs the signing in DSM.
+func makeDsmSigner(client *sdkms.Client, name string, keyHash uint32, key *sdkms.SobjectDescriptor) *dsmSigner {
+	return &dsmSigner{
+		name: name,
+		hash: keyHash,
+		client: client,
+		key: key,
+	}
+}
+
+func getDsmKeys() (note.Verifier, note.Signer, error) {
+	if len(*dsmApiKey) == 0 {
+		return nil, nil, fmt.Errorf("When signing with DSM, --dsm_key_id and --dsm_api_key must both be provided")
```

Comment:
It is always better to define a diagnostic struct and return it rather than multiple vars.